core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,243 @@
+use std::{
+    fmt::Debug,
+    sync::{Arc, Mutex},
+};
+
+use anyhow::Context;
+use async_trait::async_trait;
+use tokio::{runtime::Handle, sync::watch};
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_state::{PostgresStorage, ReadStorage, RocksdbStorage, StateKeeperColumnFamily};
+use zksync_storage::RocksDB;
+
+/// Encapsulates a storage that can produce a [`ReadStorageFactory`] on demand.
+///
+/// This struct's main design purpose is to be able to produce a (cheap) owned [`ReadStorageFactory`]
+/// implementation that can then be used to obtain a short-lived [`ReadStorage`] handle bound to
+/// the respective factory.
+#[derive(Debug, Clone)]
+pub struct StateKeeperStorage<T: ReadStorageFactory> {
+    // FIXME: `Arc<Mutex<_>>` is a bit of a leaky abstraction here as it exists mostly because of
```

Comment:
You could encapsulate `Arc<Mutex<_>>` wrapper inside `AsyncRocksdbCache` with something like

```rust
pub struct AsyncRocksdbCache(Arc<Mutex< AsyncRocksdbCacheInner>>);

enum AsyncRocksdbCacheInner {
    Postgres(ConnectionPool),
    Rocksdb(RocksDB<StateKeeperColumnFamily>, ConnectionPool),
}
```

One other thing you could try is to use [`OnceCell`](https://docs.rs/once_cell/latest/once_cell/sync/struct.OnceCell.html) to avoid a `Mutex` wrapper, so something like

```rust
struct AsyncRocksdbCache {
     pool: ConnectionPool,
     rocksdb: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
}
```